### PR TITLE
Fix arrays_zip to not rely on broken segmented gather

### LINKED ISSUE #7469: [BUG] test_arrays_zip failures on nightly
**Describe the bug**
```
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Byte)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304419.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Short)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304598.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Integer)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304777.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Long)] - py4j.protocol.Py4JJavaError: An error occurred while calling o304956.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Float)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305135.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Double)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305314.collec...
[2023-01-06T11:40:23.945Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(String)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305493.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Boolean)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305672.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Date)] - py4j.protocol.Py4JJavaError: An error occurred while calling o305851.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Timestamp)] - py4j.protocol.Py4JJavaError: An error occurred while calling o306038.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Null)] - py4j.protocol.Py4JJavaError: An error occurred while calling o306217.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Decimal(7,3))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306396.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Decimal(12,2))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306575.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Decimal(20,2))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306754.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Array(Short))] - py4j.protocol.Py4JJavaError: An error occurred while calling o306933.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Array(String))] - py4j.protocol.Py4JJavaError: An error occurred while calling o307112.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Struct(['child0', Byte],['child1', String],['child2', Float]))] - py4j.protocol.Py4JJavaError: An error occurred while calling o307291.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Map(String(not_null),String))] - py4j.protocol.Py4JJavaError: An error occurred while calling o307470.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip[Array(Binary)] - py4j.protocol.Py4JJavaError: An error occurred while calling o307649.collec...
[2023-01-06T11:40:23.946Z] FAILED ../../src/main/python/array_test.py::test_arrays_zip_corner_cases - py4j.protocol.Py4JJavaError: An error occurred while calling o307828.collec...
[2023-01-06T11:40:23.946Z] = 20 failed, 15074 passed, 830 skipped, 572 xfailed, 235 xpassed, 911 warnings in 6812.06s (1:53:32) =
```

Errors:
```
[2023-01-06T11:40:23.930Z] E                   	at java.lang.Thread.run(Thread.java:750)

[2023-01-06T11:40:23.930Z] E                   Caused by: java.lang.IllegalArgumentException: All columns must have the same number of rows

[2023-01-06T11:40:23.930Z] E                   	at ai.rapids.cudf.ColumnView.makeStructView(ColumnView.java:3440)

[2023-01-06T11:40:23.930Z] E                   	at ai.rapids.cudf.ColumnView.makeStructView(ColumnView.java:3458)

[2023-01-06T11:40:23.930Z] E                   	at ai.rapids.cudf.ColumnVector.makeStruct(ColumnVector.java:395)

[2023-01-06T11:40:23.930Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.$anonfun$zipArrays$2(collectionOperations.scala:825)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:46)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:44)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.withResource(collectionOperations.scala:701)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.$anonfun$zipArrays$1(collectionOperations.scala:824)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.withResource(collectionOperations.scala:701)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.zipArrays(collectionOperations.scala:823)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.$anonfun$columnarEval$5(collectionOperations.scala:782)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:46)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:44)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.withResource(collectionOperations.scala:701)

[2023-01-06T11:40:23.931Z] E                   	at org.apache.spark.sql.rapids.GpuArraysZip.columnarEval(collectionOperations.scala:732)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.GpuAlias.columnarEval(namedExpressions.scala:109)

[2023-01-06T11:40:23.931Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)

```

### PR DESCRIPTION
**Pull request by revans2**
This fixes #7469

Sorry that the patch is so large. I had to fundamentally change how some parts of this work and then did some refactoring to clean it up.

Before the code relied on segmented gather returning a child value for every gather entry even if the input column was null. then they would hack a null back on the output, possibly creating an invalid output where a row was marked as NULL but had a non-null offset.

This changed it so that we clean up the inputs so that the validity is the same everywhere. This should cause the children output from the segmented gather to all have the exact same offsets, so we can pull them apart properly.  This is inefficient, because we have to make a copy of all the input if any of them have a NULL. But I didn't want to risk passing an invalid column to segmented gather and having it produce an incorrect result.

I plan on filing another issue for us to stop using  `mergeAndSetValidity` on any nested type. This includes strings.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
I filed https://github.com/NVIDIA/spark-rapids/issues/7485 as follow on work for this.

**Comment by ttnghia** [comment]
I didn't read this PR code carefully (will do later). It seems that cudf's `segmented_gather` does not enforce to produce zero length segments for nulls (https://github.com/rapidsai/cudf/blob/22087b3c2a78b49140fdf96091b5abd325427b11/cpp/src/lists/copying/segmented_gather.cu#L105-L111). Should fixing it (to always produce zero length segments for nulls) will help improve performance in this PR?

**Comment by ttnghia** [comment]
I started to read the code in this PR and wonder if this is a bug or something we need to change in `segmented_gather` instead.

**Comment by revans2** [comment]
@ttnghia you have it backwards. The old version would not return a zero length segment for NULL. Your patch to make_list https://github.com/rapidsai/cudf/pull/12370 fixed it and now it does the right thing and produce a zero length segment. I don't know if it is guaranteed to always do that or if https://github.com/rapidsai/cudf/pull/12370 is fixing it up afterwards.  Either way this patch is dealing with that new reality.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala [hunk 1]
```diff
@@ -729,62 +729,66 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
...
+   * size for all of the children of the input columns. The result of this will have the same
+   * validity for all top level columns, but possibly different offsets.
+   */
+  private def normalizeNulls(inputs: Seq[cudf.ColumnVector]): Seq[ColumnVector] = {
+    // First lest figure out if there are any nulls at all, because if there are not we don't
+    // need to do anything.
+    if (inputs.exists(_.hasNulls)) {
+      var nullOutput = inputs.head.isNull
+      try {
+        inputs.slice(1, inputs.length).foreach { cv =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala:761]:
```suggestion
        inputs.drop(1).foreach { cv =>
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala [hunk 2]
```diff
@@ -809,28 +813,65 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
 
   private def generateSeqIndices(maxArraySize: ColumnVector): ColumnVector = {
     withResource(GpuScalar.from(0, IntegerType)) { s =>
-      val zeroCV = GpuColumnVector.from(s, maxArraySize.getRowCount.toInt, IntegerType)
-      withResource(zeroCV.getBase) { zero =>
+      withResource(cudf.ColumnVector.fromScalar(s, maxArraySize.getRowCount.toInt)) { zero =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala:816]: Nit: cudf's ColumnVector is already imported and is being used directly one line below.
```suggestion
      withResource(ColumnVector.fromScalar(s, maxArraySize.getRowCount.toInt)) { zero =>
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala [hunk 3]
```diff
@@ -809,28 +813,65 @@ case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with Sh
...
-      zipped.mergeAndSetValidity(BinaryOp.BITWISE_AND, inputs: _*)
+  }
+
+  /**
+   * This turns LIST[X], LIST[Y], ... into a LIST[STRUCT[X, Y, ...]] but requires that
+   * the input LIST columns all have the same validity and offsets.
+   */
+  private def zipArrays(padded: Seq[ColumnVector]) : ColumnVector = {
+    // Get the data column from the children, without any offsets
+    val children = padded.safeMap(_.getChildColumnView(0))
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala:860]: These child column views are leaked

**Review by jlowe — APPROVED** [review]

**Pull request merged**